### ISSUE TITLE
Device Switching Node - Feel free to comment or integrate

### ISSUE THREAD

**Issue by Duemellon** [pollockjj/ComfyUI-MultiGPU#179]
I had AI create this node. Don't ask me for support. If it doesn't work for you or you want other features as Gemini or ChatGPT. I really have no idea how any of this works just that it does work (to a degree)

I use this when I can't use multigpu to specify (like with loading LTX VAE using KJ nodes) or when there's problems getting some generated item to the proper GPU (like generating an image, VAE encoding, then using a sampler that's on a different GPU).

**USAGE GUIDE**
The Core Concept: "The Bridge" 

In a standard ComfyUI workflow, everything usually happens on the default GPU (cuda:0). If you run out of VRAM, the generation fails. 

This node acts as a bridge between hardware. It allows you to split a heavy workflow across multiple graphics cards. 

     GPU 0 (cuda:0): Handles the heavy lifting (Model loading, Sampling).
     GPU 1 (cuda:1): Handles the post-processing (VAE Decoding, Image Saving).
     

Workflow 1: The "Split Decode" (Most Common Use Case) 

Best for: Generating high-resolution images or large batches that fit in memory for sampling, but crash when decoding the VAE. 

In this scenario, we generate the Latent data on the first GPU and decode it into an image on the second GPU. 
Step-by-Step Setup: 

    Load Model (GPU 0) 
         Add a Load Checkpoint node.
         Implicitly: ComfyUI loads this to your default GPU (usually cuda:0).
         Connect the MODEL and VAE outputs.
          

    Prompt & Sample (GPU 0) 
         Set up your CLIP Text Encode and KSampler nodes as normal.
         Let the KSampler run on cuda:0. It outputs a LATENT.
          

    The Switch Node (The Transfer) 
         Add the Unified Device Switch (Duemellon) node.
         Set target_device: Select cuda:1 (or your secondary GPU).
         Connect latent: Connect the LATENT output from KSampler here.
         Connect vae: Connect the VAE output from the Checkpoint Loader here. This is crucial! You must move the VAE model to the second GPU so it can process the data there.
          

    Decode Image (GPU 1) 
         Add a VAE Decode node.
         Connect latent: From the latent output of the Switch Node.
         Connect vae: From the vae output of the Switch Node.
         Result: The VAE Decode will now run on cuda:1, using the second GPU's VRAM.
          

    Save Image 
         Connect the IMAGE output from VAE Decode to Save Image.
          

Result: GPU 0 is free to start the next generation immediately while GPU 1 handles the image decoding. 
Workflow 2: Model Offloading 

Best for: Using multiple distinct models in a sequence (e.g., upscale or img2img) without unloading the first model. 

If you want to run a second pass (like an upscale or a detailer) on a second GPU using a different model: 

    First Pass (GPU 0) 
         Load Model A. Generate Image A.
         Add a Unified Device Switch node.
         Set target to cuda:1.
         Pass the Image through the switch.
          

    Second Pass (GPU 1) 
         Add a second Unified Device Switch node at the start of your second pass.
         Set target to cuda:1.
         Load Model B (Connect to this Switch's model input).
         Connect the Image from the first pass.
         Note: By connecting Model B to the Switch node with target cuda:1, you force the second model to load onto the second GPU.
          

Important Notes & Tips 
1. Moving the VAE is Critical 

If you only move the Latent to cuda:1 but leave the VAE input empty or connected to a VAE still on cuda:0, the VAE Decode node will likely crash or run extremely slowly because it has to constantly copy data back and forth between GPUs. Always move both the Data (Latent) and the Processor (VAE) to the same target device. 
2. The debug_info Output 

Connect a Show Text (from Primitive nodes) or simply hover over the output to see the debug_info. 

     Example Output: Target: cuda:1 | Moved Tensors: 12 | Skipped: 0
     If "Moved Tensors" is 0, your data was already on that device, and nothing happened.
     

3. Performance Penalty 

Moving data between GPUs takes time (PCIe bandwidth). 

     Don't use this node randomly between every step.
     Do use it only when transitioning between major stages (e.g., after Sampling, before VAE Decoding).
     

4. Handling "WanVideo" 

If you are using WanVideo models (which are very large), this node is specifically designed to help distribute that workload. You can keep the heavy Video Model on cuda:0 and offload the resulting embeds or text processing to cuda:1 using the dedicated wan_model and wan_text_embeds inputs. 

**DETAILED DESCRIPTION**
This is a ComfyUI custom node designed for multi-GPU or memory management workflows. 

Its primary function is to act as a "traffic controller" that forcefully moves various types of data (images, latents) and heavy models (Checkpoints, VAEs) from one hardware device to another (e.g., from GPU 0 to GPU 1, or from GPU to CPU). 
What it does (Summary) 

In simple terms: It ensures that whatever you plug into it exists on the specific GPU or CPU you select. 

If you have a complex workflow where you might generate a latent on GPU 0 but want to decode it using a VAE on GPU 1, standard ComfyUI nodes might crash because the data is on the wrong device. This node bridges that gap. 
Detailed Features 

1. Universal Device Transfer
It supports almost every major data type used in ComfyUI, meaning you can route your entire workflow through this single node to change devices. It handles: 

     Latents (Your generation data)
     Conditioning (Positive and Negative prompts)
     Images & Masks
     Models (Standard Diffusion models)
     VAEs (Variational Autoencoders)
     WanVideo Types (Specific embeddings and models for WanVideo workflows)
     

2. Solves Multi-GPU Problems
If you have multiple graphics cards, this node allows you to orchestrate "Pipeline Parallelism." 

     Example: You can load a model on cuda:0, generate a latent, pass it through this node to move it to cuda:1, and then run a VAE decode on the second card.
     

3. Recursive Deep Search
The code uses a recursive_move function. ComfyUI data structures are often nested dictionaries (a list inside a dict inside a dict). This node doesn't just look at the top layer; it digs deep into the data structure to find every single Tensor (data block) and moves it to the target device. 

4. Model & VAE Moving
Unlike nodes that only move image data, this node calls .to(device) on the actual Model and VAE objects. This effectively offloads the model weights from one VRAM to another (or to system RAM). 
How it works (Technical Breakdown) 

    Device Selection: It populates a dropdown list of available devices (cpu, cuda:0, cuda:1, mps for Mac). 
    Execution: When executed, it iterates through every connected input. 
    Tensors: If it finds a Tensor (data) on the wrong device, it moves it. 
    Models: If it finds a Model object, it attempts to call the model's .to() method to relocate the model weights. 
    Passthrough: It passes the data out to the next node, unchanged except for its device location. 
    Debugging: It prints a log and outputs a string showing how many tensors were moved vs. skipped. 

<img width="1252" height="1102" alt="Image" src="https://github.com/user-attachments/assets/8912c804-6dbd-4acc-9b6b-f2c6501316e4" />

```
import torch
import time

print(">>> LOADING UNIFIED DEVICE SWITCH MODULE (Duemellon) - VAE SUPPORT <<<")

class UnifiedDeviceSwitch:
    """
    A centralized bridge for multi-GPU workflows.
    Moves Data AND Models to a specific target device.
    Includes Dual Conditioning passthroughs for Positive/Negative prompts and VAE support.
    """
    
    CATEGORY = "Duemellon/Utils"
    FUNCTION = "move_data"
    RETURN_TYPES = ("LATENT", "CONDITIONING", "CONDITIONING", "IMAGE", "MASK", "WANVIDIMAGE_EMBEDS", "WANVIDEOTEXTEMBEDS", "WANVIDEOMODEL", "MODEL", "VAE", "STRING")
    RETURN_NAMES = ("latent", "conditioning_positive", "conditioning_negative", "image", "mask", "wan_img_embeds", "wan_text_embeds", "wan_model", "model", "vae", "debug_info")

    def __init__(self):
        print("[Duemellon Switch] Node Instantiated.")

    @classmethod
    def INPUT_TYPES(cls):
        device_list = ["cpu"]
        if torch.cuda.is_available():
            count = torch.cuda.device_count()
            for i in range(count):
                device_list.append(f"cuda:{i}")
        if hasattr(torch.backends, "mps") and torch.backends.mps.is_available():
             device_list.append("mps")

        return {
            "required": {
                "target_device": (device_list, {
                    "default": "cpu", 
                    "tooltip": "The hardware device (e.g., cuda:1) where the data needs to be moved to."
                }),
            },
            "optional": {
                "latent": ("LATENT",),
                "conditioning_positive": ("CONDITIONING", {
                    "tooltip": "Connect Positive Conditioning here."
                }),
                "conditioning_negative": ("CONDITIONING", {
                    "tooltip": "Connect Negative Conditioning here."
                }),
                "image": ("IMAGE",),
                "mask": ("MASK",),
                "wan_img_embeds": ("WANVIDIMAGE_EMBEDS",),
                "wan_text_embeds": ("WANVIDEOTEXTEMBEDS",),
                "wan_model": ("WANVIDEOMODEL", {
                    "tooltip": "Connect the WanVideo Model here to force-move it to the target GPU."
                }),
                "model": ("MODEL", {
                    "tooltip": "Connect standard Models here."
                }),
                "vae": ("VAE", {
                    "tooltip": "Connect VAE here to move it to the target GPU."
                }),
            }
        }

    @classmethod
    def IS_CHANGED(cls, **kwargs):
        return float(time.time())

    def move_data(self, target_device, latent=None, conditioning_positive=None, conditioning_negative=None, image=None, mask=None, wan_img_embeds=None, wan_text_embeds=None, wan_model=None, model=None, vae=None):
        device = torch.device(target_device)
        
        print(f"\n[Duemellon Switch] EXECUTING move_data. Target: {device}")
        
        stats = {"moved": 0, "skipped": 0, "types": set()}

        def recursive_move(obj, key_name=None, depth=0):
            indent = "  " * depth

            if key_name == "MODEL_MOVE" and depth == 0:
                print(f"{indent}Attempting to move Model/Wrapper/VAE of type: {type(obj)}")

            if key_name == "generator_state":
                return obj

            if isinstance(obj, torch.Tensor):
                stats["types"].add(str(obj.dtype))
                if str(obj.device) != str(device):
                    stats["moved"] += 1
                    return obj.to(device)
                else:
                    stats["skipped"] += 1
                    return obj

            elif isinstance(obj, list):
                return[recursive_move(x, key_name, depth+1) for x in obj]
            elif isinstance(obj, tuple):
                return tuple(recursive_move(x, key_name, depth+1) for x in obj)
            elif isinstance(obj, dict):
                return {k: recursive_move(v, key_name=k, depth=depth+1) for k, v in obj.items()}
            
            elif hasattr(obj, 'to') and callable(obj.to):
                try:
                    if key_name == "MODEL_MOVE":
                         print(f"{indent}  -> Calling .to({device}) on object...")
                    
                    return obj.to(device)
                except Exception as e:
                    print(f"{indent}  -> FAILED to move object: {e}")
                    pass

            elif hasattr(obj, '__iter__') and not isinstance(obj, (str, bytes)):
                 try:
                     return[recursive_move(x, key_name, depth+1) for x in obj]
                 except Exception:
                     pass

            return obj

        out_latent = recursive_move(latent) if latent is not None else None
        out_cond_pos = recursive_move(conditioning_positive) if conditioning_positive is not None else None
        out_cond_neg = recursive_move(conditioning_negative) if conditioning_negative is not None else None
        out_image = recursive_move(image) if image is not None else None
        out_mask = recursive_move(mask) if mask is not None else None
        out_wan_img = recursive_move(wan_img_embeds) if wan_img_embeds is not None else None
        out_wan_text = recursive_move(wan_text_embeds) if wan_text_embeds is not None else None
        
        out_wan_model = recursive_move(wan_model, key_name="MODEL_MOVE") if wan_model is not None else None
        out_std_model = recursive_move(model, key_name="MODEL_MOVE") if model is not None else None
        out_vae = recursive_move(vae, key_name="MODEL_MOVE") if vae is not None else None

        debug_str = f"Target: {target_device} | Moved Tensors: {stats['moved']} | Skipped: {stats['skipped']}"
        print(f"[Duemellon Switch] {debug_str}")
        
        return (out_latent, out_cond_pos, out_cond_neg, out_image, out_mask, out_wan_img, out_wan_text, out_wan_model, out_std_model, out_vae, debug_str)

NODE_CLASS_MAPPINGS = {
    "UnifiedDeviceSwitch": UnifiedDeviceSwitch
}
NODE_DISPLAY_NAME_MAPPINGS = {
    "UnifiedDeviceSwitch": "Unified Device Switch (Duemellon)"
}
```

**Comment by Duemellon** [comment]
Created newer, better version. IT's better b/c there's only one input & you choose the type from dropdown so it takes up less visual space. However, it still doesn't resolve the LTX23 problem for my workflow.

```
import torch
import time

class AnyType(str):
    def __eq__(self, _):
        return True
    def __ne__(self, _):
        return False

any_type = AnyType("*")

def get_device_list():
    device_list = ["cpu"]
    if torch.cuda.is_available():
        count = torch.cuda.device_count()
        for i in range(count):
            device_list.append(f"cuda:{i}")
    if hasattr(torch.backends, "mps") and torch.backends.mps.is_available():
         device_list.append("mps")
    return device_list

class UnifiedDeviceSwitch:
    """
    Moves data to a specified target device.
    Utilizes a dropdown for visual identification of the connected data type.
    """
    CATEGORY = "Duemellon/Utils"
    FUNCTION = "move_data"
    RETURN_TYPES = (any_type, "STRING")
    RETURN_NAMES = ("data", "debug_info")

    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "target_device": (get_device_list(), {
                    "default": "cpu", 
                    "tooltip": "The hardware device to move the data to."
                }),
                "expected_type": ([
                    "LATENT", 
                    "CONDITIONING", 
                    "IMAGE", 
                    "MASK", 
                    "MODEL", 
                    "VAE", 
                    "WANVIDIMAGE_EMBEDS", 
                    "WANVIDEOTEXTEMBEDS", 
                    "WANVIDEOMODEL"
                ], {
                    "tooltip": "Select the type of data being connected for visual context."
                }),
                "data": (any_type, {
                    "tooltip": "Connect the data payload here."
                }),
            }
        }

    @classmethod
    def IS_CHANGED(cls, **kwargs):
        return float(time.time())

    def move_data(self, target_device, expected_type, data):
        device = torch.device(target_device)
        
        stats = {"moved": 0, "skipped": 0, "failed": 0, "types": set()}

        def recursive_move(obj, key_name=None):
            if key_name == "generator_state":
                if isinstance(obj, torch.Tensor):
                    return obj.to("cpu")
                return obj

            if isinstance(obj, torch.Tensor):
                stats["types"].add(str(obj.dtype))
                if str(obj.device) != str(device):
                    stats["moved"] += 1
                    return obj.to(device)
                else:
                    stats["skipped"] += 1
                    return obj

            elif isinstance(obj, list):
                return[recursive_move(x) for x in obj]
            elif isinstance(obj, tuple):
                return tuple(recursive_move(x) for x in obj)
            elif isinstance(obj, dict):
                return {k: recursive_move(v, key_name=k) for k, v in obj.items()}
            
            elif hasattr(obj, 'to') and callable(obj.to):
                stats["types"].add(type(obj).__name__)
                try:
                    return obj.to(device)
                except Exception as e:
                    stats["failed"] += 1
                    return obj

            elif hasattr(obj, '__iter__') and not isinstance(obj, (str, bytes)):
                 try:
                     return [recursive_move(x) for x in obj]
                 except Exception:
                     pass

            return obj

        out_data = recursive_move(data)
        
        debug_str = f"Target: {target_device} | Expected: {expected_type} | Moved: {stats['moved']} | Skipped: {stats['skipped']} | Failed: {stats['failed']} | Types: {list(stats['types'])}"
        print(f"[Duemellon Switch] {debug_str}")
        
        return (out_data, debug_str)

class UnifiedSamplerSwitch:
    """
    Synchronizes KSampler inputs to a single target device.
    """
    CATEGORY = "Duemellon/Utils"
    FUNCTION = "sync_sampler_data"
    RETURN_TYPES = ("MODEL", "CONDITIONING", "CONDITIONING", "LATENT", "STRING")
    RETURN_NAMES = ("model", "positive", "negative", "latent_image", "debug_info")

    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "target_device": (get_device_list(), {
                    "default": "cpu",
                    "tooltip": "The hardware device to move the sampler components to."
                }),
                "model": ("MODEL",),
                "positive": ("CONDITIONING",),
                "negative": ("CONDITIONING",),
                "latent_image": ("LATENT",),
            }
        }

    @classmethod
    def IS_CHANGED(cls, **kwargs):
        return float(time.time())

    def sync_sampler_data(self, target_device, model, positive, negative, latent_image):
        device = torch.device(target_device)
        stats = {"moved": 0, "skipped": 0, "failed": 0, "types": set()}

        def recursive_move(obj, key_name=None):
            if key_name == "generator_state":
                if isinstance(obj, torch.Tensor):
                    return obj.to("cpu")
                return obj

            if isinstance(obj, torch.Tensor):
                stats["types"].add(str(obj.dtype))
                if str(obj.device) != str(device):
                    stats["moved"] += 1
                    return obj.to(device)
                else:
                    stats["skipped"] += 1
                    return obj

            elif isinstance(obj, list):
                return [recursive_move(x) for x in obj]
            elif isinstance(obj, tuple):
                return tuple(recursive_move(x) for x in obj)
            elif isinstance(obj, dict):
                return {k: recursive_move(v, key_name=k) for k, v in obj.items()}
            
            elif hasattr(obj, 'to') and callable(obj.to):
                stats["types"].add(type(obj).__name__)
                try:
                    return obj.to(device)
                except Exception as e:
                    stats["failed"] += 1
                    return obj

            elif hasattr(obj, '__iter__') and not isinstance(obj, (str, bytes)):
                 try:
                     return [recursive_move(x) for x in obj]
                 except Exception:
                     pass

            return obj

        out_model = recursive_move(model)
        out_pos = recursive_move(positive)
        out_neg = recursive_move(negative)
        out_latent = recursive_move(latent_image)

        debug_str = f"Target: {target_device} | Sync Completed | Moved: {stats['moved']} | Skipped: {stats['skipped']} | Failed: {stats['failed']}"
        print(f"[Duemellon Sampler Sync] {debug_str}")
        
        return (out_model, out_pos, out_neg, out_latent, debug_str)
```
